shared/search/providers.ts

```diff
@@ -249,23 +238,154 @@ function getFileContent(
+    const searchResults = await search(query)
+    const preFilteredResults = searchResults.filter(result =>
+        isSymbolVisible(doc, path, result)
+    )
+
+    // Filter results based on language spec
+    const filteredResults = filterDefinitions(preFilteredResults, {
+        repo,
+        filePath: path,
+        fileContent: text,
+    })
+
+    return sortByProximity(
+        filteredResults.map(resultToLocation),
+        new URL(doc.uri)
+    )
+}
+
+/**
+ * Perform a search query for references. Returns results converted to locations.
+ * Results are not sorted in any meaningful way as these results are meant to be
+ * merged with other search query results.
+ *
+ * @param args Parameter bag.
+ */
+async function searchReferences({
+    query,
+}: {
+    /** The search query. */
+    query: string
+}): Promise<sourcegraph.Location[]> {
+    return (await search(query)).map(resultToLocation)
+}
+
+/**
+ * Invoke the given search function by modifying the query with a term that will
+ * only look in the current git tree by appending a repo filter with the repo name
+ * and the current commit.
+ *
+ * This is likely to timeout on large repos or organizations with monorepos if the
+ * current commit is not an indexed commit. Instead of waiting for a timeout, we
+ * will start a second index-only search of the HEAD commit for the same repo after
+ * a short delay.
+ *
+ * This function returns the set of results that resolve first.
+ *
+ * @param search The search function.
+ * @param args The arguments to the search function.
+ */
+export function searchWithFallback<
+    P extends { repo: string; commit: string; query: string },
+    R
+>(search: (args: P) => Promise<R>, args: P): Promise<R> {
+    const { repo, commit, query } = args
+    const unindexedQuery = `${query} repo:^${repo}$@${commit}`
+    const indexedQuery = `${query} repo:^${repo}$ index:only`
+
+    if (getConfig('basicCodeIntel.indexOnly', false)) {
+        return search({
+            ...args,
+            query: indexedQuery,
+        })
+    }
+
+    const timeout = getConfig<number>(
+        'basicCodeIntel.unindexedSearchTimeout',
+        5000
+    )
+
+    return raceWithDelayOffset(
+        search({ ...args, query: unindexedQuery }),
+        () => search({ ...args, query: indexedQuery }),
+        timeout
+    )
+}
+
 /**
  * Perform a search query.
  *
  * @param query The search query.
  */
 async function search(query: string): Promise<Result[]> {
-    if (shouldTraceSearch()) {
+    if (getConfig('basicCodeIntel.debug.traceSearch', false)) {
         console.log('%cSearch', 'font-weight:bold;', {
             query,
         })
     }
 
-    return (await searchViaApi(query, shouldRequestFileLocal())).flatMap(
+    return (await searchViaApi(query, getConfig('fileLocal', false))).flatMap(
         searchResultToResults
     )
 }
 
+/**
+ * Filter search results before passing them to the language spec filter function.
```

Comment:
This comment is partly pre-existing, but I think we can improve it by talking about what the function does rather than how it's used. What I _think_ this means is something like:

```typescript
// Report whether the given symbol search result is visible to the document
// with the given text and file path. A symbol is visible if it is heuristically
// public or is defined in the file named by path.
```

I am not completely sure of this wording (e.g., is it "defined" or "located"? Is it "visible TO" or "visible FROM"?) but I'd like to focus the comment more on what it means for `isSymbolVisible(doc, path, sym)` to be true.

Reply:
I think it can be even simpler.

```
Filter out search results that are a non-public symbol that does not belong to the current text document.
```

Reply:
I agree that's simpler, but it only addresses part of what I was hoping to convey:

1. This function does not filter anything, it is a predicate. I was hoping to get the comment to describe what the predicate _means_, rather than what it's used for.
2. The name seems not to match the semantics: It's not really about visibility so much as location. Maybe something like `isExternalPrivateSymbol` or something.

Reply:
Ah I see!

Pushed an update. Merging now, but could you check that one last time for correctness?